Move log_prefill_stats_late to correct location in PP mode

harvenstar opened the pull request:
Address review feedback from #15943.

  Move `log_prefill_stats_late()` call from `prefill.py:526` to `scheduler_pp_mixin.py:224` (right after `get_new_batch_prefill()`).

  The original location was incorrect because `event_loop_overlap_disagg_prefill()` doesn't create new batches. The new location matches the calling pattern in non-PP mode (`prefill.py:334`).

<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.io to discuss further. -->

## Motivation

<!-- Describe the purpose and goals of this pull request. -->

## Modifications

<!-- Detail the changes made in this pull request. -->

## Accuracy Tests

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->

## Benchmarking and Profiling

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

ShangmingCai reviewed (commented):
Why there are many fixes of this `log_prefill_stats_late` recently? I am seeing like 3 PRs are addressing this problem. If this is need for `event_loop_pp_disagg_prefill`, should `event_loop_pp` also add this line?

harvenstar commented:
> > @ShangmingCai Why there are many fixes of this log_prefill_stats_late recently? I am seeing like 3 PRs are addressing this problem. If this is need for event_loop_pp_disagg_prefill, should event_loop_pp also add this line?
> 
> 1. I introduced that for a useful metric in DP attention mode, but as is suggested by @hnyls2002 I do not run CI but use CI on main to save resources
> 2. thus the CI expectedly fails in several places and thus there are fixes to this
> 3. `If this is need for event_loop_pp_disagg_prefill, should event_loop_pp also add this line?` - generally speaking every `.log_prefill_stats` (called inside `get_new_batch_prefill`) should be paired with one `.log_prefill_stats_late` after prepare_mlp_sync_batch is called, since the "late" function exists b/c I need information obtained in prepare_mlp_sync_batch for metrics, thus yes I guess so. cc @harvenstar .
>    
>    * Another way may be somehow add a maybe_prepare_mlp_sync_batch and include this, based on how many places we changed (not realize this when I firstly impl it since lack of CI) this may be reasonable.

@fzyzcjy Thanks for the clarification. I took a look at the calling patterns:

event_loop_pp (around line 85):
self.mbs[mb_id] = self.get_next_batch_to_run()
get_next_batch_to_run() already pairs log_prefill_stats (from get_new_batch_prefill()) with log_prefill_stats_late() (after prepare_mlp_sync_batch()), so I suppose no change is needed there.

event_loop_pp_disagg_prefill (around line 220-224):

batch = self.get_new_batch_prefill()  # log_prefill_stats called here
if self.require_mlp_sync:
    batch = self.prepare_mlp_sync_batch(batch)
self.log_prefill_stats_late(batch)    # this PR adds the pairing


This path directly calls the lower-level functions, so it needs the manual pairing. This PR fixes that.

So event_loop_pp is already correct. The missing pairing was only in event_loop_pp_disagg_prefill (and get_next_disagg_prefill_batch_to_run, which was addressed in #15943).

On the refactoring suggestion, I am happy to add a helper like maybe_prepare_mlp_sync_batch_and_log_stats() to reduce duplication. Would you prefer that change to be included in this PR, or should I send it as a small follow-up PR to keep this one minimal?

fzyzcjy commented:
> @fzyzcjy: Another way may be somehow add a maybe_prepare_mlp_sync_batch and include this, based on how many places we changed (not realize this when I firstly impl it since lack of CI) this may be reasonable.
> @harvenstar: Would you prefer that change to be included in this PR, or should I send it as a small follow-up PR to keep this one minimal?

I guess just refactor it here is enough

harvenstar committed:
Refactor: unify all prepare_mlp_sync_batch calls with helper function

Add `maybe_prepare_mlp_sync_batch_and_log_stats()` helper in SchedulerDPAttnMixin
to ensure log_prefill_stats (called in get_new_batch_prefill) is always paired
with log_prefill_stats_late after prepare_mlp_sync_batch.

Changes:
- Add helper with optional need_sync and log_stats parameters
- Replace all Scheduler-level prepare_mlp_sync_batch calls:
  * event_loop_pp_disagg_prefill (scheduler_pp_mixin.py)
  * get_next_disagg_prefill_batch_to_run (prefill.py)
  * get_next_disagg_decode_batch_to_run (decode.py)
  * get_next_batch_to_run (scheduler.py, both calls at 1831 and 1846)

This unifies the pairing pattern across all event loops and reduces duplication.

fzyzcjy commented:
shall we change all, instead of some, of the calls

<img width="1328" height="716" alt="image" src="https://github.com/user-attachments/assets/4623c637-ae97-4c13-bfe5-2c15c48d809d" />

harvenstar commented:
> shall we change all, instead of some, of the calls
> 
> 
> 
> <img width="1328" height="716" alt="image" src="https://github.com/user-attachments/assets/4623c637-ae97-4c13-bfe5-2c15c48d809d" />
> 
> 

sure!

ShangmingCai reviewed (approved):
Thx. The logic looks very clear now.

ShangmingCai commented:
/tag-and-rerun-ci

ShangmingCai commented:
cc: @ch-wan

ch-wan reviewed (approved):
LGTM

fzyzcjy reviewed (approved):
LGTM, and we should add back the temporarily removed assertion again once everything is good

fzyzcjy commented:
I guess ensure not break ci then everything is ok (but here is dilemma - we do not want to use precious ci resource on this pr but w/ that we are hard to verify it is correct...)
